ctapipe/ml/sklearn.py

```diff
@@ -0,0 +1,193 @@
+"""
+Component Wrappers around sklearn models
+"""
+
+import joblib
+import numpy as np
+import astropy.units as u
+from traitlets import Bool, Dict, Enum, Integer, List, Unicode
+
+from sklearn.utils import all_estimators
+
+from ..core import Component, Provenance
+from .preprocessing import check_valid_rows, table_to_float
+
+SUPPORTED_CLASSIFIERS = dict(all_estimators("classifier"))
+SUPPORTED_REGRESSORS = dict(all_estimators("regressor"))
+SUPPORTED_MODELS = {**SUPPORTED_CLASSIFIERS, **SUPPORTED_REGRESSORS}
+
+
+class Model(Component):
+    features = List(Unicode(), help="Features to use for this model").tag(config=True)
+    model_config = Dict({}, help="kwargs for the sklearn model").tag(config=True)
+    model_cls = Enum(SUPPORTED_MODELS.keys(), default_value=None, allow_none=False).tag(
+        config=True
+    )
+    target = Unicode(None, allow_none=False, help="Name of target column").tag(
+        config=True
+    )
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        # to verify settings
+        self.new_model()
+
+        self.models = {}
+        self.unit = None
+
+    def new_model(self):
+        return SUPPORTED_MODELS[self.model_cls](**self.model_config)
+
+    def table_to_X(self, table):
+        feature_table = table[self.features]
+        valid = check_valid_rows(feature_table, log=self.log)
+        X = table_to_float(feature_table[valid])
+        return X, valid
+
+    def table_to_y(self, table, mask=None):
+        if self.unit is not None:
+            return table[mask][self.target].quantity.to_value(self.unit)
+        return np.array(table[self.target][mask])
+
+    def fit(self, key, table):
+        self.models[key] = self.new_model()
+
+        X, valid = self.table_to_X(table)
+        y = self.table_to_y(table, mask=valid)
+        self.unit = table[self.target].unit
+        self.models[key].fit(X, y)
+
+    def predict(self, key, table):
+        if key not in self.models:
+            raise KeyError(
+                f"No model available for key {key},"
+                f" available models: {self.models.keys()}"
+            )
+        X, valid = self.table_to_X(table)
+        n_outputs = getattr(self.models[key], "n_outputs_", 1)
+
+        if n_outputs > 1:
+            shape = (len(table), n_outputs)
+        else:
+            shape = (len(table),)
+
+        prediction = np.full(shape, np.nan)
+        if np.any(valid):
+            prediction[valid] = self.models[key].predict(X)
+
+        if self.unit is not None:
+            prediction = u.Quantity(prediction, self.unit, copy=False)
+
+        return prediction, valid
+
+    def write(self, path):
+        Provenance().add_output_file(path, role="ml-model")
+        with open(path, "wb") as f:
+            joblib.dump(self, f, compress=True)
+
+    @classmethod
+    def load(cls, path, check_cls=True):
```

Comment:
When is this load function supposed to be used? If I run `ctapipe-ml-train-regessor` and run `Regressor.load("output.pkl.gz")` on it's output, it fails, as this pkl file seems to have the regressor and a few other objects in it.    I don't see anywhere where this function is even used, but maybe I missed it?

Reply:
It looks like Reconstructor.read() does it's own IO, so maybe we should remove this here? or else Reconstructor.load() should use this function.

Reply:
In the refactoring I completely removed the classes formerly in `sklearn.py`, IO now only happens through pickling the reconstructors.